test/unit/ngsiv2/amqpBinding-test.js

```diff
@@ -0,0 +1,320 @@
+            // Note: /v1/updateContext response is not processed by IOTA so its content is irrelevant,
+            // as far as it is a 200 OK
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v1/updateContext')
+                .reply(200, '{}');
+
+
+            contextBrokerMock
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v2/entities/SensorMachine:UL_UNPROVISIONED/attrs',
+                 utils.readExampleFile('./test/unit/ngsiv2/contextRequests/unprovisionedMeasure.json'))
+                .reply(204);
+
+            request(groupCreation, function(error, response, body) {
+                done();
+            });
+        });
+
+        it('should send a new update context request to the Context Broker with just that attribute', function(done) {
+            channel.publish(config.amqp.exchange, '.80K09H324HV8732.UL_UNPROVISIONED.attrs.temperature',
+                new Buffer('23'));
+
+            setTimeout(function() {
+                contextBrokerMock.done();
+                done();
+            }, 100);
+        });
+    });
+
+    describe('When a new multiple measure arrives to a Device routing key with one measure', function() {
+        beforeEach(function() {
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities/Second%20UL%20Device/attrs',
+                    utils.readExampleFile('./test/unit/ngsiv2/contextRequests/singleMeasure.json'))
+                .reply(204);
+        });
+
+        it('should send a single update context request with all the attributes', function(done) {
+            channel.publish(config.amqp.exchange, '.1234.MQTT_2.attrs', new Buffer('temperature|23'));
+
+            setTimeout(function() {
+                contextBrokerMock.done();
+                done();
+            }, 100);
+        });
+    });
+
+    describe('When a new multiple measure arrives to a Device routing key with a faulty payload', function() {
+        beforeEach(function() {
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities/Second%20UL%20Device/attrs',
+                    utils.readExampleFile('./test/unit/ngsiv2/contextRequests/singleMeasure.json'))
+                .reply(204);
+        });
+
+        it('should silently ignore the error (without crashing)', function(done) {
+            channel.publish(config.amqp.exchange, '.1234.MQTT_2.attrs', new Buffer('notAULPayload '));
+
+            setTimeout(function() {
+                done();
```

Comment:
Maybe some error condition should be checked here? Something printed in the log (although I'm not sure if it is easy to intercept logs from a unit test...)

Reply:
Hi, 
In this tests, the agent work  like the following: Internaly, it request for path [/iot/devices], 
```
msg=Request for path [/iot/devices] from [localhost:4041] | comp=IoTAgent
time=2018-03-23T09:46:00.038Z | lvl=DEBUG | corr=8bf40727-6a5f-4a53-9cad-b8e2ba0e7b32 | trans=8bf40727-6a5f-4a53-9cad-b8e2ba0e7b32 | op=IoTAgentNGSI.GenericMiddlewares | srv=smartGondor | subsrv=/gardens | msg=Body:

{
    "devices": [
        {
            "protocol": "GENERIC_PROTO",
            "transport": "AMQP",
            "device_id": "MQTT_2",
            "entity_name": "Second UL Device",
            "entity_type": "AnMQTTDevice",
            "timezone": "America/Santiago",
            "attributes": [
                {
                    "name": "temperature",
                    "type": "celsius"
                },
                {
                    "name": "humidity",
                    "type": "degrees"
                }
            ]
        }
    ]
}


```
then it looks for entity params ["device","subsevice","type"] in order to register the device into NGSI:

```
 msg=Looking for entity params ["service","subservice","type"] | comp=IoTAgent
time=2018-03-23T09:46:00.082Z | lvl=DEBUG | corr=8bf40727-6a5f-4a53-9cad-b8e2ba0e7b32 | trans=8bf40727-6a5f-4a53-9cad-b8e2ba0e7b32 | op=IoTAgentNGSI.DeviceService | srv=smartGondor | subsrv=/gardens | msg=Registering device into NGSI Service:
{
    "id": "MQTT_2",
    "type": "AnMQTTDevice",
    "name": "Second UL Device",
    "service": "smartGondor",
    "subservice": "/gardens",
    "active": [
        {
            "name": "temperature",
            "type": "celsius",
            "object_id": "temperature"
        },
        {
            "name": "humidity",
            "type": "degrees",
            "object_id": "humidity"
        }
    ],
    "staticAttributes": [],
    "lazy": null,
    "commands": [],
    "timezone": "America/Santiago",
    "protocol": "GENERIC_PROTO",
    "transport": "AMQP",
    "internalId": null,
    "subscriptions": []
}

```
Once is registered, no context provider registrations has been found, but the entity is created:

```
 {
    "url": "http://192.168.1.1:1026/v1/updateContext",
    "method": "POST",
    "json": {
        "contextElements": [
            {
                "type": "AnMQTTDevice",
                "isPattern": "false",
                "id": "Second UL Device",
                "attributes": [
                    {
                        "name": "temperature",
                        "type": "celsius",
                        "value": " "
                    },
                    {
                        "name": "humidity",
                        "type": "degrees",
                        "value": " "
                    }
                ]
            }
        ],
        "updateAction": "APPEND"
    },
    "headers": {
        "fiware-service": "smartGondor",
        "fiware-servicepath": "/gardens",
        "fiware-correlator": "8bf40727-6a5f-4a53-9cad-b8e2ba0e7b32"
    }
}


```

Reply:
In the last version of the code in this part I see:

```
            setTimeout(function() {
            //FIXME/TODO
                done();
            }, 100);
        });
    });
```

Please, provide extra information after the FIXME/TODO mark (and chose one of both), e.g:

```
// FIXME: nothing to check in this case except expectations in log file. However, currently we doen't have a mechanism to check expentations in log files, so this marked for future improvement.
```

or a similar workding expressing the same idea.

(In addition, it seems there is an indent issue, i.e. the `//` should be aligned with the `done()` line before).

Reply:
Fixed in d7d186654283f4b33ad3f445b6c9a99bb50ac6ac